### DISCUSSION TITLE
Sending a OTP and Magic Link

### DISCUSSION THREAD

**Discussion by Jet315** [supabase/supabase#19806] [Questions]
Hello,

I am trying to send both a Magic link (so a user can click to sign in) as well as a 6 digit OTP (User can also enter this to verify signin/signup)

I am not actually sure if this is possible. I am able to send both a ConfirmationURL and Token in the email template, but the Token never seems to be valid, I am using https://supabase.com/docs/reference/python/auth-verifyotp in order to verify the 6 digit token.

Any support or whether this is doable would be great.

Code to send magic link + OTP:  (Magic link works fine,  6 digit token is also sent)
```python
 data = supabase.auth.sign_in_with_otp({
            "email": '<EMAIL>',
            "options": {
                "email_redirect_to": 'https://example.com/welcome'
            }
        })
``` 

Code to verify 6 digit token:

```python
res = supabase.auth.verify_otp ({
                "email": event["email"],
                "token": token_value,
                "type": "magiclink" //I have also tried signup and email
            })
``` 

Error I get is:
`[ERROR] AuthApiError: Token has expired or is invalid`

Thanks,

**Comment by Cyberniee** [comment]
Same here, did you figure out a solution?
I worked around it by verifying the jwt separately, but it does not entirely align with the "supabase simplicity" I'd say

**Reply by Jet315** [reply]
Nope didn't find a solution, for the time being I plan just to use the Magic Link. It's odd that the Magic Link email template on Supabase allows you to include the 6 digit token for it then not to be valid.

**Reply by sujjeee** [reply]
Hey, I'm using the Supabase magic link feature but keep getting the same error `AuthApiError: Email link is invalid or has expired. `
Did you find out any fix for this? I'm using the service key for creating the magic link.

```
const { data, error } = await supabaseAdmin.generateLink({
  type: "magiclink",
  email: email,
  options: {
    redirectTo: `${env.NEXT_PUBLIC_APP_URL}/auth/confirm`,
  },
})
```

It seems to be working fine for the first time, but when creating a new magic link, it's generating the same token. I suspect this is why it's showing the error.
